demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,288 @@
+                                            nameIndex,
+                                            filterIndex,
+                                            &status );
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            matchFound = matchWildcards( pTopicFilter,
+                                         pTopicName,
+                                         topicNameLength,
+                                         filterIndex,
+                                         &nameIndex,
+                                         &status );
+        }
+
+        if( matchFound == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( status == false )
+    {
+        /* If the end of both strings has been reached, they match. */
+        status = ( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength,
+                        bool requiresExactMatch )
+{
+    bool status = false;
+
+    /* Check for an exact match if the incoming topic name and the registered
+     * topic filter length match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 );
+    }
+    else
+    {
+        status = topicFilterMatch( pTopicName, topicNameLength, pTopicFilter, topicFilterLength );
+    }
+
+    return status;
+}
+
+/* Design -
+ *   * Common handler can contain logic of processing acks and retries.
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( uint16_t packetIdentifier,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( packetIdentifier != MQTT_PACKET_ID_INVALID );
+
+    size_t listIndex = 0u;
+
+    /* Iterate through record list to find matching topics, and invoke their callbacks. */
+    while( listIndex < recordListCount )
+    {
+        if( matchTopic( pPublishInfo->pTopicName,
+                        pPublishInfo->topicNameLength,
+                        callbackRecordList[ listIndex ].pTopicFilter,
+                        callbackRecordList[ listIndex ].topicFilterLength,
+                        callbackRecordList[ listIndex ].requiresExactMatch ) == true )
+        {
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex++ ].callback( packetIdentifier, pPublishInfo );
```

Comment:
I think this will result in an infinite loop if matchTopic returns false

Reply:
Good catch! Fixed